### ISSUE TITLE
Temperature spikes occurring frequently

### ISSUE THREAD

**Issue by louisaw123** [openmsupply/msupply-cold-chain#251]
**Describe the bug**
Temperature spikes occurring in fridge monitoring sensors frequently. Extremely high temperatures occurring that range 60 degrees and above. Spike temps recorded over 3 months:

![Screenshot 2023-10-20 at 2 59 21 PM](https://github.com/openmsupply/msupply-cold-chain/assets/143057869/d1c1ffaf-a29c-4f6b-a691-98130df6d09e)

Spike temperature is not accurate as other sensors are also kept near the affected sensor but show normal readings. Examples:

![Screenshot 2023-10-20 at 3 02 10 PM](https://github.com/openmsupply/msupply-cold-chain/assets/143057869/8f391d33-8fe9-4b26-ae3c-9a8ec145b54d)

![Screenshot 2023-10-20 at 3 03 15 PM](https://github.com/openmsupply/msupply-cold-chain/assets/143057869/f93d7dde-bbd7-4613-aa4a-0c6afa3d8ff2)

**To Reproduce**
Unknown

**Expected behavior**
Temperature should be reading accurately otherwise may cause inaccurate data analysis. It also makes it difficult for user data reliability particularly for those monitoring without a backup sensor. 

**Additional context**
The issue is not isolated to a single site and has been noticed in other sites too. Blue Maestro have said that they have not come across this issue on their end and suspect it might be something caused on our end - have been monitoring own sensors with CCA and tempo but have not come across any spikes yet to confirm this.

**Comment by adamdewey** [comment]
Weird that they're all in a similar temp range...

**Comment by adamdewey** [comment]
Not had any spikes in the last 90 days for any of the sensors on the vaccine server 🤔 

<img width="1512" alt="image" src="https://github.com/openmsupply/msupply-cold-chain/assets/19318937/190a18dd-d056-44ea-9642-9b06049001b6">

**Comment by adamdewey** [comment]
I wonder if it could be caused by older sensors? 

I think client could have sensors that are 2+ years old, hmm

**Comment by jmbrunskill** [comment]
@adamdewey  - This reminds me that we might want some kind of filter for this in notify? Should we ignore temperatures that are over or under certain limits? We don't want people getting alerts saying their fridge is 70 degrees we it's actually fine...

**Comment by adamdewey** [comment]
Hey @jmbrunskill - yeah I agree, wouldn't want to alert for those fake spikes. 

I think that could be a good idea to implement even before hiding the spikes in CCA, Open mSupply etc as alerts would be highly visible.

**Comment by adamdewey** [comment]
@jmbrunskill - do we need a separate issue in the Notify repo, or can we link multiple PRs from different repos to this issue?

**Comment by louisaw123** [comment]
umm hold your horses... i don't think you should filter any temps? if its high it might be high for a legitimate reason.. i have no idea what the alert plans are but are they not going to get an alert only after having lets say 15-20 mins of high readings - rather than immediately? if that is the case then a spike will not affect anything...

**Comment by jmbrunskill** [comment]
> umm hold your horses... i don't think you should filter any temps? if its high it might be high for a legitimate reason.. i have no idea what the alert plans are but are they not going to get an alert only after having lets say 15-20 mins of high readings - rather than immediately? if that is the case then a spike will not affect anything...

The current plan it to alert as soon as it goes high so someone can go and check why, e.g. if someone left to door open, the power's off so they need to start a generator or something.
The idea is to allow someone to respond quickly before there is a `breach` which would be when the temperature had been too high or two low for a long time.

**Comment by louisaw123** [comment]
hmm.. what is the temp considered "breach" here because temps can jump above 8 pretty quickly and then jump back down like from opening a fridge door to transfer stock. sometimes having too many alerts might make people desensitised.

usually fridges have an alarm that beeps if it happened in the first instance and you would generally try and do something if you were already there... 

just weighing up instant notification vs notifying after x mins option anyway

**Comment by adamdewey** [comment]
yeah good points Louisa...

Maybe we need to think about it some more - i.e. how can we help them to prevent a breach happening in the first place?

**Comment by adamdewey** [comment]
> Hey I passed this msg on to Andrei last week but have not had a chance to catch up with u guys about it. Had a small sensor temp blip last week. We assume it's more related to the sensor? and also assumed bcos it was running with AA alkaline..? Although technically did not look into the cause in depth. Just thought I'd keep u guys updated too

![telegram-cloud-photo-size-1-5091601867787578386-y](https://github.com/msupply-foundation/msupply-cold-chain/assets/19318937/7264f7b3-5f9e-4cea-b22f-3b321653f7f0)

That last one is more worrying as it's a realistic temperature so harder to filter out

**Comment by adamdewey** [comment]
Couple more spikes: 

![image](https://github.com/msupply-foundation/msupply-cold-chain/assets/19318937/4d9bcc06-9671-4488-a87b-fc2b1d6e3eb7)

**Comment by louisaw123** [comment]
top one is a spike the bottom one is not - can see the datapoints gradually going up and back down

**Comment by adamdewey** [comment]
> top one is a spike the bottom one is not - can see the datapoints gradually going up and back down

Ah good spotting LDP!

**Comment by adamdewey** [comment]
This spike occurred while using the new AA battery cases

![image](https://github.com/msupply-foundation/msupply-cold-chain/assets/19318937/7b703f29-eeda-45d9-8def-4a1404816a25)

**Comment by adamdewey** [comment]
For the client above spikes have been seen on 3 out of 19 sensors (so far)

For another client they've been seen on 4 out of 12 sensors

**Comment by adamdewey** [comment]
Temporary fix will be to mask high outliers (above 55 C) in both the dashboard and notification service. 

Full fix is likely to be to go with BLE packet advertising - added to backlog

**Comment by marthakeezy** [comment]
Just noticed another spike (in case you guys need more cases)
<img width="1100" alt="Screenshot 2024-01-30 at 9 04 21 AM" src="https://github.com/msupply-foundation/msupply-cold-chain/assets/153777484/87361bb7-1c1a-4f01-993b-a8b15eb9fc7e">

**Comment by louisaw123** [comment]
![image](https://github.com/msupply-foundation/msupply-cold-chain/assets/143057869/7ae75e05-cde8-421b-8981-0585027c1e1f)

20.4 degree spike

**Comment by louisaw123** [comment]
and 57.6 

![image](https://github.com/msupply-foundation/msupply-cold-chain/assets/143057869/0ed38cf6-0f0b-4dce-a065-e628aea2f819)

**Comment by louisaw123** [comment]
<img width="1233" alt="Screenshot 2024-02-26 at 1 41 33 PM" src="https://github.com/msupply-foundation/msupply-cold-chain/assets/143057869/bda68974-2bf7-4cce-9c15-14b57fd7470a">

**Comment by adamdewey** [comment]
Check out this for some weird spike behaviour:

![image](https://github.com/msupply-foundation/msupply-cold-chain/assets/19318937/c803baf2-aa03-46d0-aba7-2a711f51c652)

Multiple super high spikes in a row

**Comment by louisaw123** [comment]
Experiment with finding BM Spikes:
------------------------------------------
METHOD: 

LOL im gonna forget this ill write it:

1. take sensor. install it into cca app
2. turn off cca app
3. after 3 weeks, open BM app and connect to each sensor , download logs
4. open cca app and sync sensors
5. Compare data from BM app and CCA to see if any difference

(Notes re method: the sensors i use were all previously connected to my test tablet already at home so i am just disconnecting it from this tablet and installing it into the cca app on a new tablet for this test)

------------------------------------------
RESULTS:

Sensor 1: EA56ABB1
![Screenshot 2024-03-21 at 2 41 02 PM](https://github.com/msupply-foundation/msupply-cold-chain/assets/143057869/29721729-4c0d-4678-aebd-ec04a693c80f)

Sensor 2: CFE0C3B5
![Screenshot 2024-03-21 at 3 24 26 PM](https://github.com/msupply-foundation/msupply-cold-chain/assets/143057869/5eebe733-4df4-4c89-a034-30f8c009fd96)

Sensor 3: F143B0EC
![Screenshot 2024-03-21 at 3 25 41 PM](https://github.com/msupply-foundation/msupply-cold-chain/assets/143057869/1dfa026e-d4ea-4a00-9369-474aebfbf25a)


-------------------------------------------
OBSERVATIONS:

1) Note that temperatures on the CCA App are recording 10 minutes behind from Blue Maestro app when comparing on spreadsheet. This is due to the tablet I am using having the incorrect time (it is 10 minutes behind from real time) and the Blue Maestro data being pulled from my phone which is on real time. So if a time on the CCA app data says 20:15, it is technically meant to be 20:25.

2) the orange table shows data from blue maestro app while blue table shows data from cold chain app. i have aligned all records next to each other, they all appear fine except for the first record

3) for sensor 1 @ 20:25 -> sensor read at 15.6 on cca while on bm it reads at 6.6...... sensor did not leave fridge during this time and can see this in the bm app from the records before this. With sensor 1 there are also records from 7pm which could have been due to the sensor not syncing to the old tablet hence extra records sitting in there. interestingly these records are picked up on bm app but are not picked up on cca app?

4) sensor 2 and sensor 3 both have different temperatures on cca app vs bm app. no other temps were different. 

Have added the excel files for the rest of the records...
For BM app: sensor 1 recorded 4628 records, while sensor 2 and 3 recorded 4614 records.

[F143B0EC_1710921030505.xls](https://github.com/msupply-foundation/msupply-cold-chain/files/14681699/F143B0EC_1710921030505.xls)

[EA56ABB1_1710921337407.xls](https://github.com/msupply-foundation/msupply-cold-chain/files/14681698/EA56ABB1_1710921337407.xls)


[CFE0C3B5_1710921108745.xls](https://github.com/msupply-foundation/msupply-cold-chain/files/14681697/CFE0C3B5_1710921108745.xls)

**Comment by adamdewey** [comment]
I tried to repro this (spike on first reading) with 6000 logs for 3 sensors with a 1 min log interval (~3 days of data) but couldn't get the same result unfortunately

<img width="748" alt="image" src="https://github.com/msupply-foundation/msupply-cold-chain/assets/19318937/c8d0c5c3-3ebd-4aaf-bcfa-92dde37c2cbb">

<img width="779" alt="image" src="https://github.com/msupply-foundation/msupply-cold-chain/assets/19318937/eec8f5e2-ef4e-4258-9c5e-bba48d00fb03">

<img width="778" alt="image" src="https://github.com/msupply-foundation/msupply-cold-chain/assets/19318937/ae1166ad-1eb4-4193-b5c1-92cd6f50a5e0">

**Comment by jmbrunskill** [comment]
I think I might have seen a temperature spike on the blue maestro app today wasn't fast enough to screenshot it though. Maybe I confused a humidity reading...

**Comment by adamdewey** [comment]
hmm, you mean it showed and then you think they masked it after? 

Because the historical data should be accessible on the blue maestro app?

**Comment by jmbrunskill** [comment]
Nah I just saw it on the home page of their app. Can try downloading logs...

**Comment by adamdewey** [comment]
Ooh, that would be interesting if you did manage to catch one!

**Comment by adamdewey** [comment]
Double spike seen on a client just now: 

<img width="1500" alt="image" src="https://github.com/msupply-foundation/msupply-cold-chain/assets/19318937/0c89ba7a-7a7d-422e-bce1-3289eb7ef415">

**Comment by adamdewey** [comment]
@louisaw123 caught a spike using the new 0.5.7 version. 

Upon investigation it appears that CCA is occasionally taking the humidity reading instead of the temperature reading (😱)

Spike noticed on dashboard at 27.7 °C:

<img width="1384" alt="image" src="https://github.com/user-attachments/assets/3ba6b23f-9152-409a-9f5f-ac444a381b91">


Spike seen on CCA at 27.7 °C:

<img width="640" alt="image" src="https://github.com/user-attachments/assets/449efd64-d6db-45d8-8d7c-48901d5572e5">

Sensor logs showing humidity reading of 27.7 °C:

<img width="1030" alt="image" src="https://github.com/user-attachments/assets/ed1ee293-878a-49de-81f0-325bdfc8737f">

Export of sensor logs: 
[C1671B0D_1730231726948_log.csv](https://github.com/user-attachments/files/17564525/C1671B0D_1730231726948_log.csv)

Export of CCA logs: 
[C1671B0D64C9-30-10-2024-09-33-53.csv](https://github.com/user-attachments/files/17564528/C1671B0D64C9-30-10-2024-09-33-53.csv)

Debug logs for CCA: 
[com.msupplycoldchain-latest.3.log](https://github.com/user-attachments/files/17564531/com.msupplycoldchain-latest.3.log)
[com.msupplycoldchain-2024-10-29.14.log](https://github.com/user-attachments/files/17564532/com.msupplycoldchain-2024-10-29.14.log)
[com.msupplycoldchain-2024-10-29.15.log](https://github.com/user-attachments/files/17564533/com.msupplycoldchain-2024-10-29.15.log)
[com.msupplycoldchain-2024-10-29.16.log](https://github.com/user-attachments/files/17564534/com.msupplycoldchain-2024-10-29.16.log)
[com.msupplycoldchain-2024-10-29.17.log](https://github.com/user-attachments/files/17564530/com.msupplycoldchain-2024-10-29.17.log)

**Comment by adamdewey** [comment]
Captured a negative spike today, the spike doesn't exist on the sensor logs, only in CCA. 

![image](https://github.com/user-attachments/assets/8fbfdd4c-6409-4ec4-a313-40343b100c41)

Possibly has taken the dewpoint reading instead of temperature, although the logs don't line up exactly: 

<img width="671" alt="image" src="https://github.com/user-attachments/assets/9d4235d1-abde-4b62-954f-c5fba052c7ae">

[E1185AAF693B-31-10-2024-10-11-51.csv](https://github.com/user-attachments/files/17579284/E1185AAF693B-31-10-2024-10-11-51.csv)
[E1185AAF_1730322650968_log.csv](https://github.com/user-attachments/files/17579286/E1185AAF_1730322650968_log.csv)

**Comment by kahn-msupply** [comment]
Another spike recorded. v0.5.6.

![image](https://github.com/user-attachments/assets/ad38234b-9b69-4e6c-8e8e-05df610c1329)
